src/main/java/org/takes/misc/VerboseList.java

```diff
@@ -0,0 +1,205 @@
+
+    /**
+     * Error message for IndexOutOfBoundsException.
+     */
+    private final transient String message;
+
+    /**
+     * Ctor.
+     * @param list Original list
+     * @param msg Error message for IndexOutOfBoundsException
+     */
+    public VerboseList(final List<T> list, final String msg) {
+        this.origin = list;
+        this.message = msg;
+    }
+
+    @Override
+    public int size() {
+        return this.origin.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return this.origin.isEmpty();
+    }
+
+    @Override
+    public boolean contains(final Object obj) {
+        return this.origin.contains(obj);
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        return new VerboseIterator<T>(this.origin.iterator(), this.message);
+    }
+
+    @Override
+    public Object[] toArray() {
+        return this.origin.toArray();
+    }
+
+    @SuppressWarnings("PMD.UseVarargs")
+    @Override
+    public <E> E[] toArray(final E[] arr) {
+        return this.origin.toArray(arr);
+    }
+
+    @Override
+    public boolean add(final T item) {
+        return this.origin.add(item);
+    }
+
+    @Override
+    public boolean remove(final Object obj) {
+        return this.origin.remove(obj);
+    }
+
+    @Override
+    public boolean containsAll(final Collection<?> coll) {
+        return this.origin.containsAll(coll);
+    }
+
+    @Override
+    public boolean addAll(final Collection<? extends T> coll) {
+        return this.origin.addAll(coll);
+    }
+
+    @Override
+    public boolean addAll(final int index, final Collection<? extends T> coll) {
+        try {
+            return this.origin.addAll(index, coll);
+        } catch (final IndexOutOfBoundsException ex) {
+            throw new IllegalArgumentException(this.message, ex);
+        }
+    }
+
+    @Override
+    public boolean removeAll(final Collection<?> coll) {
+        return this.origin.removeAll(coll);
+    }
+
+    @Override
+    public boolean retainAll(final Collection<?> coll) {
+        return this.origin.retainAll(coll);
+    }
+
+    @Override
+    public void clear() {
+        this.origin.clear();
+    }
+
+    @Override
+    public T get(final int index) {
+        try {
+            return this.origin.get(index);
+        } catch (final IndexOutOfBoundsException ex) {
+            throw new IllegalArgumentException(this.message, ex);
+        }
+    }
+
+    @Override
+    public T set(final int index, final T element) {
+        try {
+            return this.origin.set(index, element);
+        } catch (final IndexOutOfBoundsException ex) {
+            throw new IllegalArgumentException(this.message, ex);
+        }
+    }
+
+    @Override
+    public void add(final int index, final T element) {
+        try {
+            this.origin.add(index, element);
+        } catch (final IndexOutOfBoundsException ex) {
+            throw new IllegalArgumentException(this.message, ex);
```

Comment:
@Happy-Neko Here too. No need to wrap the exception, just let it be rethrown since it's unchecked.